### PR TITLE
PRO-1035 Token Management

### PR DESCRIPTION
**Pull request by hemadri-tank-bacancy**
https://linear.app/pillarproject/issue/PRO-1035/pillar-wallet-token-management

## Description

- Implemented, Tokens in settings.
- Implemented, Tokens button under wallet to open token management
- Implemented, filter by chain and tokenlists to view the tokens that are already added by default pillar tokenlist. 
- Implemented, tap on a tokenlist to see the tokens and which ones are already added.
- Implemented, how many tokenlists a token is present. 
- Implemented, enable and disable any token from viewing them in the wallet.
- Implemented, search for a specific token by name and ticker after tapping on a tokenlist.
- Implemented, import contract addresses and add the token to the wallet if they are not present on a tokenlist. It will pull the token logo and data for me.
- Implemented, if enabled a token, it must be present in the wallet view under All, Tokens.
- Implemented, if I have enabled, I can pull the balance from the token and the token price. 
- Implemented, If I have enabled a new token or added it, this token must be available in the swap selection in To so I can buy it, this should be the default list showing all tokens + added tokens locally.

## Motivation and Context

- 

## How Has This Been Tested?

- Checked Import custom token, Add tokens list, Custom tokens in assets, merged tokens list in assets
- Tested in locally device Samsung A50 (Mainnet + Testnets)
- Tested in Simulator iPhone 13 (Mainnet + Testnets)
-

## Screenshots (if appropriate):
Introduction:
https://user-images.githubusercontent.com/78720269/226516368-ad8f0a15-446c-48ff-aea8-eef4fc08a487.mp4

Add Token:
https://user-images.githubusercontent.com/78720269/226517071-6bd42c53-673f-42aa-8bc0-234dbfe50201.mp4

Manage Custom Token:
https://user-images.githubusercontent.com/78720269/226518104-9260a556-37f5-4f07-a14d-2ca2bbacc35e.mp4

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: src/utils/chains.js [hunk 1]
```diff
@@ -54,6 +54,38 @@ export const chainFromChainId: { [number]: Chain } = {
...
+  CHAIN_ID.GOERLI,
+  CHAIN_ID.BINANCE_TESTNET,
+  CHAIN_ID.SOKOL,
+  CHAIN_ID.MUMBAI,
+  CHAIN_ID.FUJI,
+  CHAIN_ID.OPTIMISM_GOERLI,
+  CHAIN_ID.ARBITRUM_NITRO,
+];
+
+const MainNetsChainIds = [
```
Comment (Yash-Chaubal) [inline src/utils/chains.js:67]:
```suggestion
const mainnetChainIds = [
```

File: src/utils/chains.js [hunk 2]
```diff
@@ -54,6 +54,38 @@ export const chainFromChainId: { [number]: Chain } = {
...
+  CHAIN_ID.AVALANCHE,
+  CHAIN_ID.OPTIMISM,
+  CHAIN_ID.ARBITRUM,
+];
+
+export const isTestNetsChainId = (chainId: number) => {
+  const isAvilChainId = TestNetsChainIds.some((id) => chainId === id);
+  if (isAvilChainId) return true;
+  return false;
+};
```
Comment (Yash-Chaubal) [inline src/utils/chains.js:81]:
```suggestion
export const isTestnetChainId = (chainId: number) => testnetChainIds.some((id) => chainId === id);
```

File: src/utils/chains.js [hunk 3]
```diff
@@ -54,6 +54,38 @@ export const chainFromChainId: { [number]: Chain } = {
...
+
+export const isMainNetsChainId = (chainId: number) => {
+  const isAvilChainId = MainNetsChainIds.some((id) => chainId === id);
+  if (isAvilChainId) return true;
+  return false;
+};
```
Comment (Yash-Chaubal) [inline src/utils/chains.js:87]:
```suggestion
export const isMainnetChainId = (chainId: number) => mainnetChainIds.some((id) => chainId === id);
```

**Review by Yash-Chaubal — CHANGES_REQUESTED** [review]
Looking good so far! Just added some comments to consider before merging

File: src/actions/assetsActions.js [hunk 4]
```diff
@@ -698,18 +714,16 @@ export const fetchSupportedAssetsAction = () => {
...
 
-        const popularAssets = isEmpty(updatedPopularAssets?.[chain]) ? [] : updatedPopularAssets?.[chain];
-
-        const removedDuplicateSupportedAssets = chainSupportedAssets.filter(
+        const removedDuplicateSupportedAssets = chainCustomAssets.filter(
           (item) =>
-            !popularAssets?.some(
-              (popularAsset) => item.symbol === popularAsset?.symbol && item.address === popularAsset?.address,
+            !chainSupportedAssets?.some(
+              (customAsset) => item.symbol === customAsset?.symbol && item.address === customAsset?.address,
```
[the remark this answers is not in the document]
Comment (hemadri-tank-bacancy) [reply src/actions/assetsActions.js]: For multiple chain comparison in need both address and symbol. updated some other changes

File: src/actions/assetsActions.js [hunk 5]
```diff
@@ -799,3 +782,113 @@ export const localAssets = (chain: Chain) => {
...
+
+    if (isEmpty(tokenList)) {
+      reportErrorLog('assetsActions addTokensListAction: fetching remote config data failed', {
+        key: REMOTE_CONFIG.APP_TOKENLISTS,
+      });
+      dispatch({ type: ADD_TOKENS_FETCHING, payload: false });
+      return;
+    }
+
+    const parsedTokenLists = JSON.parse(tokenList);
```
[the remark this answers is not in the document]
Comment (hemadri-tank-bacancy) [reply src/actions/assetsActions.js]: I updated it

File: src/actions/assetsActions.js [hunk 6]
```diff
@@ -799,3 +782,113 @@ export const localAssets = (chain: Chain) => {
...
+    } = getState();
+
+    const isMainnet = isProdEnv();
+
+    // nothing to do if offline
+    if (!isOnline || isFetching || isFetched) return;
+
+    dispatch({ type: ADD_TOKENS_FETCHING, payload: true });
+
+    const tokenList = firebaseRemoteConfig.getString(REMOTE_CONFIG.APP_TOKENLISTS);
```
[the remark this answers is not in the document]
Comment (hemadri-tank-bacancy) [reply src/actions/assetsActions.js]: I updated it

File: src/utils/assets.js [hunk 7]
```diff
@@ -409,3 +421,60 @@ export const sortInvestmentPositions = (positionsInfo: AccountInvestmentPosition
   });
   return newArr;
 };
+
+export const isSame = (a: Asset, b: Asset) => a.symbol === b.symbol && a.address === b.address;
```
[the remark this answers is not in the document]
Comment (hemadri-tank-bacancy) [reply src/utils/assets.js]: I updated it

File: src/utils/chains.js [hunk 8]
```diff
@@ -54,6 +54,38 @@ export const chainFromChainId: { [number]: Chain } = {
   [CHAIN_ID.ARBITRUM_NITRO]: CHAIN.ARBITRUM,
 };
 
+const TestNetsChainIds = [
```
[the remark this answers is not in the document]
Comment (hemadri-tank-bacancy) [reply src/utils/chains.js]: I updated it

File: src/utils/chains.js (hunk 1 above)
Comment (hemadri-tank-bacancy) [reply src/utils/chains.js]: I updated it

File: src/utils/chains.js (hunk 2 above)
Comment (hemadri-tank-bacancy) [reply src/utils/chains.js]: I updated it

File: src/utils/chains.js (hunk 3 above)
Comment (hemadri-tank-bacancy) [reply src/utils/chains.js]: I updated it

File: src/actions/assetsActions.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Yash-Chaubal) [reply src/actions/assetsActions.js]: The symbol could be made up by the user, and on line 717 we're already filtering `chainSupportedAssets` so its for a specific chain - the address is the only identifier for a unique asset

File: src/utils/chains.js [hunk 9]
```diff
@@ -54,6 +54,30 @@ export const chainFromChainId: { [number]: Chain } = {
...
+  CHAIN_ID.XDAI,
+  CHAIN_ID.POLYGON,
+  CHAIN_ID.AVALANCHE,
+  CHAIN_ID.OPTIMISM,
+  CHAIN_ID.ARBITRUM,
+];
+
+export const isTestNetsChainId = (chainId: number) => testnetChainIds.some((id) => chainId === id);
+
+export const isMainNetsChainId = (chainId: number) => mainNetChainIds.some((id) => chainId === id);
```
Comment (Yash-Chaubal) [inline src/utils/chains.js:79]: The names of these vars still need to be changed
```suggestion
const mainnetChainIds = [
  CHAIN_ID.ETHEREUM_MAINNET,
  CHAIN_ID.BINANCE,
  CHAIN_ID.XDAI,
  CHAIN_ID.POLYGON,
  CHAIN_ID.AVALANCHE,
  CHAIN_ID.OPTIMISM,
  CHAIN_ID.ARBITRUM,
];

export const isTestnetChainId = (chainId: number) => testnetChainIds.some((id) => chainId === id);

export const isMainnetChainId = (chainId: number) => mainnetChainIds.some((id) => chainId === id);
```

**Review by Yash-Chaubal — COMMENTED** [review]
Still some var names to change, and i've added another comment to the thread about comparing assets to consider

File: src/utils/chains.js (hunk 9 above)
Comment (hemadri-tank-bacancy) [reply src/utils/chains.js]: I updated it.

**Review by Yash-Chaubal — APPROVED** [review]
Thanks for the changes, LGTM!

**Pull request merged**